llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -3154,7 +3186,11 @@ Value *SPIRVToLLVM::indexDescPtr(Type *elementTy, Value *base, Value *index) {
   // A sampler pointer is represented by a {pointer,stride,convertingSamplerIdx} struct. If the converting sampler
   // index is non-zero (i.e. it is actually a converting sampler), we also want to modify that index. That can only
   // happen if there are any converting samplers at all.
-  if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3) {
+  Type *pointerTy = base->getType()->getStructElementType(0);
+  unsigned vecCount = (pointerTy->getNumContainedTypes() != 0)
+                          ? cast<FixedVectorType>(pointerTy->getPointerElementType())->getNumElements()
+                          : 0;
+  if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3 && vecCount == 4) {
```

Comment:
I don't understand why this change is necessary. Also, I don't think it will work with opaque pointers.

Reply:
There would be overlapped when translate load sampler and image descriptor. These codes are used to distinguish sampler and image descriptors loading. 

The commit could be pass Docker testings which include "amdllpc -enable-opaque-pointers=true"

Reply:
That the tests pass probably just means that there's a lack of proper test coverage for this.

Really, this code pattern simply isn't okay. In an opaque pointer world, you get `getNumContainedTypes() == 0`, which means you get `vecCount == 0`, which means this code never executes. That can't be right.

Also, I don't understand what you mean by overlap.

Reply:
The content here is used to load sampler descriptor. If only check with "if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3)", it would load image descriptor with sampler operations. Such as:

  %7 = call i32 (...) @lgc.create.get.desc.stride.i32(i32 1, i32 1, i32 1, i32 0)
  %8 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } poison, <8 x i32> addrspace(4)* %6, 0
  %9 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } %8, i32 %7, 1
  %10 = udiv i32 %7, 3
  %11 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } %9, i32 %10, 2
  %12 = insertvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } undef, { <8 x i32> addrspace(4)*, i32, i32 } %11, 0
  %13 = insertvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } %12, { <4 x i32> addrspace(4)*, i32, i32 } { <4 x i32> addrspace(4)* null, i32 0, i32 1 }, 1
  %14 = extractvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } %13, 0
  %15 = extractvalue { { <8 x i32> addrspace(4)*, i32, i32 }, { <4 x i32> addrspace(4)*, i32, i32 } } %13, 1
  **%16 = extractvalue { <8 x i32> addrspace(4)*, i32, i32 } %14, 2
  %17 = add i32 %16, 0
  %18 = icmp ne i32 %16, 0
  %19 = select i1 %18, i32 %17, i32 0
  %20 = insertvalue { <8 x i32> addrspace(4)*, i32, i32 } %14, i32 %19, 2**
  %21 = extractvalue { <8 x i32> addrspace(4)*, i32, i32 } %20, 0
  %22 = extractvalue { <8 x i32> addrspace(4)*, i32, i32 } %20, 1

And then it would go wrong into bold codes when handle following codes of SPIRVToLLVM::loadImageSampler(). Actually I'm not sure if it is a LLVM optimization latent issue.

    if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() >= 3) {
      Value *descSrdStride = getBuilder()->CreateExtractValue(base, 2);
      Type *ptrTy = ptr->getType();

      for (unsigned planeIdx = 1; planeIdx != arrayTy->getNumElements(); ++planeIdx) {
        ptr = getBuilder()->CreateBitCast(
            ptr, getBuilder()->getInt8Ty()->getPointerTo(ptr->getType()->getPointerAddressSpace()));
        ptr = getBuilder()->CreateGEP(getBuilder()->getInt8Ty(), ptr, descSrdStride);
        ptr = getBuilder()->CreateBitCast(ptr, ptrTy);
        oneVal = getBuilder()->CreateLoad(elementTy, ptr);
        result = getBuilder()->CreateInsertValue(result, oneVal, planeIdx);
      }
    }

Reply:
an alternate solution is adding two members in resource node.  {resource_ptr, stride, plane_stride, is_res}, then we only need change this condition to 
if (!m_convertingSamplers.empty() && base->getType()->getStructNumElements() == 3)

Reply:
Fixed in the latest commit, please help to review, thanks !